### PR TITLE
Implement `JOIN ... USING` query

### LINKED ISSUE #730: Support `JOIN ... USING ( col1, col2, ... )`
Many databases support this shorthand for joins when the column names are the same in both tables.

From Postgres docs:

> The `USING` clause is a shorthand that allows you to take advantage of the specific situation where both sides of the join use the same name for the joining column(s). It takes a comma-separated list of the shared column names and forms a join condition that includes an equality comparison for each one. For example, joining `T1` and `T2` with `USING (a, b)` produces the join condition `ON T1.a = T2.a AND T1.b = T2.b`.
> 
> ```
> => SELECT * FROM t1 INNER JOIN t2 ON t1.num = t2.num;
>  num | name | num | value
> -----+------+-----+-------
>    1 | a    |   1 | xxx
>    3 | c    |   3 | yyy
> (2 rows)
> 
> => SELECT * FROM t1 INNER JOIN t2 USING (num);
>  num | name | value
> -----+------+-------
>    1 | a    | xxx
>    3 | c    | yyy
> (2 rows)
> ```

PostgreSQL: https://www.postgresql.org/docs/current/queries-table-expressions.html
MySQL https://dev.mysql.com/doc/refman/8.0/en/join.html
Sqlite: https://www.sqlite.org/syntax/join-constraint.html
Oracle: https://docs.oracle.com/javadb/10.6.2.1/ref/rrefsqljusing.html
SQL Server: not supported

Usage could be something like
```ts
db.selectFrom('person as a')
  .innerJoinUsing('person as b', ['first_name', 'last_name'])
  .select(['a.id', 'b.id'])
```

### PR DESCRIPTION
**Pull request by dwickern**
Fixes #730

### CODE REVIEW

**Comment by dwickern** [comment]
To implement `JOIN ... USING`, I added a new type argument to `JoinBuilder`, `C extends string`, representing the column names which are common to both tables being joined.

I see that `JoinBuilder` is also used for the `MERGE` query. Consequently this PR allows for a `MERGE` syntax which is not valid in Postgres and probably not valid on any DBMS:
```ts
// merge into person using pet using (id)
db.mergeInto('person').using('pet', (join) => join.using(['id']))
```

There are a couple ways to solve that (if maintainers here think it's worth solving):
1. Create a `SelectJoinBuilder` which extends `JoinBuilder` and adds the `using` function. Use `SelectJoinBuilder` for `SELECT` queries and use `JoinBuilder` for `MERGE` queries
2. `MERGE` queries could assign `C = never` as the type param to `JoinBuilder`

File: src/query-builder/join-builder.ts [hunk 1]
```diff
@@ -79,6 +79,37 @@ export class JoinBuilder<DB, TB extends keyof DB>
...
+   *
+   * The generated SQL (PostgreSQL):
+   *
+   * ```sql
+   * select *
+   * from "person"
+   * inner join "pet" using ("name")
+   * ```
+   */
+  using(columns: [C, ...C[]]): JoinBuilder<DB, TB, C> {
```
Comment (dwickern) [inline src/query-builder/join-builder.ts:106]: Does it make sense to have an overload for a single column? I noticed that other APIs like `select` have this.
```js
.innerJoin('pet', join => join.using('col'))
.innerJoin('pet', join => join.using(['col1', 'col2', 'col3']))
```

File: src/operation-node/join-node.ts [hunk 2]
```diff
@@ -15,7 +16,7 @@ export interface JoinNode extends OperationNode {
   readonly kind: 'JoinNode'
   readonly joinType: JoinType
   readonly table: OperationNode
-  readonly on?: OnNode
+  readonly on?: OnNode | JoinUsingNode
```
Comment (igalklebanov) [inline src/operation-node/join-node.ts:19]: That's a breaking change not worth doing, and also a surprise - how's `on` related to `using`?

This should make things quite simple, and no need for an extra node type.

```suggestion
  readonly on?: OnNode
  readonly using?: ReadonlyArray<ColumnNode>
```

File: src/query-builder/join-builder.ts (hunk 1 above)
Comment (igalklebanov) [reply src/query-builder/join-builder.ts:106]: Seems OK to me.
I'm not sure about `[C, ...C[]]`, I don't think we're THAT strict in other places.

**Comment by igalklebanov** [comment]
> To implement `JOIN ... USING`, I added a new type argument to `JoinBuilder`, `C extends string`, representing the column names which are common to both tables being joined.

Sounds like it's not crucial and could've been placed on the method/s. We're usually quite conservative when it comes to adding more generic parameters to existing builders. There has to be a good reason..

> I see that `JoinBuilder` is also used for the `MERGE` query. Consequently this PR allows for a `MERGE` syntax which is not valid in Postgres and probably not valid on any DBMS:
> 
> ```ts
> // merge into person using pet using (id)
> db.mergeInto('person').using('pet', (join) => join.using(['id']))
> ```
> 
> There are a couple ways to solve that (if maintainers here think it's worth solving):
> 
> 1. Create a `SelectJoinBuilder` which extends `JoinBuilder` and adds the `using` function. Use `SelectJoinBuilder` for `SELECT` queries and use `JoinBuilder` for `MERGE` queries
> 2. `MERGE` queries could assign `C = never` as the type param to `JoinBuilder`

That's fine for now, not worth the extra builder/s IMHO.

**Comment by peterHakio** [comment]
@dwickern Hope you have time to finish the pull request and insure that it gets merged. Since it is a very nice feature and very nice work.

**Pull request closed without merging**